zinc/src/main/scala/sbt/internal/inc/IncrementalCompilerImpl.scala

```diff
@@ -331,18 +331,31 @@ class IncrementalCompilerImpl extends IncrementalCompiler {
         Incremental.prune(srcsSet, previousAnalysis, output, outputJarContent)
     }
 
-    // Run the incremental compilation
-    val compile = IncrementalCompile(
-      srcsSet,
-      lookup,
-      mixedCompiler.compile,
-      analysis,
-      output,
-      log,
-      incOptions,
-      outputJarContent
-    )
-    compile.swap
+    def runIncrementalCompilation(analysis: CompileAnalysis) = {
+      val compile = IncrementalCompile(
+        srcsSet,
+        lookup,
+        mixedCompiler.compile,
+        analysis,
+        output,
+        log,
+        incOptions,
+        outputJarContent
+      )
+      compile.swap
+    }
+
+    def recoverableFailure(e: CompileFailed): Boolean = {
+      e.problems.exists(_.message.contains("A full rebuild may help"))
+    }
+
+    try {
+      runIncrementalCompilation(analysis)
+    } catch {
+      case e: CompileFailed if recoverableFailure(e) =>
+        mixedCompiler.config.reporter.reset()
+        runIncrementalCompilation(Analysis.empty)
```

Comment:
I am not sure if this is a good idea. If we have false positive on the "A full rebuild may help" for whatever reason, we'd run into an infinite loop. Instead we should try to fix the invalidation for your use case.

Reply:
There shoun't be an infinite loop. This is before we are `cycling`, so if compilation fails once with this error, it will be retried only once.
I'd prefer to fix the invalidation, but to be honest I don't know how. I was thinking about the actual problem for a while and it sounds like not possible with how zinc is designed. We can't detect that prior to initial compilation, which will fail. Extending the heuristics to detect that prior to the compilation would usually  cause too many recompilations. I don't see any other way to fix this.